### PR TITLE
[WPB-5389] Guard user connection requests by team-level federation settings

### PR DESCRIPTION
**Pull request by mdimjasevic**
This PR introduces guarding of user connection requests by team-level federation settings.

Tracked by https://wearezeta.atlassian.net/browse/WPB-5389.

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

The following pre-existing 13 integration tests in Brig need to be dropped (as they've been failing due to application code changes in this PR) and rewritten in the `integration` package:
- [x] federation/POST /federation/on-user-deleted-connections : 200
- [x] post /users/connections-status/v2 : All connections
- [x] Remote connections: limits
- [x] Remote connections: send then cancel
- [x] Remote connections: block, remote cancels, then accept
- [x] Remote connections: block then accept
- [x] Remote connections: ignore, remote cancels, then accept
- [x] Remote connections: ignore then accept
- [x] Remote connections: mutual Connect - remote action then local action
- [x] Remote connections: mutual Connect - local action then remote action
- [x] post /connections (bad user)
- [x] connection/post /connections
- [x] post /clients/:cid/access-token/success

### CODE REVIEW

File: integration/test/Test/Federation.hs [hunk 1]
```diff
@@ -128,7 +128,7 @@ testNotificationsForOfflineBackends = do
 
       -- FUTUREWORK: Uncomment after fixing this bug: https://wearezeta.atlassian.net/browse/WPB-3664
       -- void $ awaitNotification downUser1 downClient1 (Just newMsgNotif) 1 isOtherUser2LeaveUpConvNotif
-      -- void $ awaitNotification otherUser otherClient (Just newMsgNotif) 1 isDelUserLeaveDownConvNotif
+      void $ awaitNotification otherUser otherClient (Just newMsgNotif) isDelUserLeaveDownConvNotif
```
Comment (mdimjasevic) [inline integration/test/Test/Federation.hs:131]: Why this change? The test is now failing. Also, it feels like this change doesn't belong to this PR.

Comment (MangoIV) [reply integration/test/Test/Federation.hs:131]: oh yeah; perhaps I can should revert it, I just looked at the PR  above and it is merged, so the test should pass and I thought I'd just give it a shot and if it doesn't pass, then I comment it again...

**Comment by MangoIV** [comment]
https://wearezeta.atlassian.net/browse/WPB-3664

**Comment by MangoIV** [comment]
https://github.com/wireapp/wire-server/pull/3591

This PR should have fixed this line (see above)

```haskell
void $ awaitNotification otherUser otherClient (Just newMsgNotif) isDelUserLeaveDownConvNotif
```

Perhaps we should have a look at it again? 

cc @pcapriotti

File: integration/test/Test/Connection.hs [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+    deleteUser charlie
+
+    -- charlie is on their local backend, so asking should be instant
+    getConnection charlie alice `bindResponse` \resp ->
+      resp.status `shouldMatchInt` 404
+
+    -- for alice, charlie is on the remote backend, so the status change
+    -- may not be instant
+    getConnection alice charlie `waitForResponse` \resp ->
+      resp.status `shouldMatchInt` 404
```
Comment (mdimjasevic) [inline integration/test/Test/Connection.hs:94]: I suppose this works too, but we used a different approach so far. In particular, Alice would wait for a notification on her web socket that Charlie got deleted. Once she has that, she is sure that `getConnection alice charlie` returns a 404.

Comment (MangoIV) [reply integration/test/Test/Connection.hs:94]: does this work if the connection is pending?

Comment (mdimjasevic) [reply integration/test/Test/Connection.hs:94]: I don't think it does as there are no notifications for this state.

I'd say it's fine to do it this way; I was just saying so far we have done it differently, but that's no argument against doing something new like you did here.

File: integration/test/Test/Connection.hs [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+  void $ putConnection alice bob "cancelled" >>= getBody 200
+  assertConnectionStatus alice bob "cancelled"
+
+testConnectionLimits :: HasCallStack => App ()
+testConnectionLimits = do
+  let connectionLimit = 16
+
+  alice <- randomUser OwnDomain def
+  [charlie1, charlie2, charlie3, charlie4] <- replicateM 4 do
```
Comment (battermann) [inline integration/test/Test/Connection.hs:245]: We've got a helper to create multiple users: `createUsers $ replicate 4 OtherDomain`

File: integration/test/Test/Connection.hs [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+    resp.status `shouldMatchInt` 201
+
+testNonFederatingRemoteTeam :: HasCallStack => App ()
+testNonFederatingRemoteTeam =
+  withFederatingBackendsAllowDynamic $ \(domainA, domainB, _) -> do
+    sequence_
+      [ createFedConn domainA (FedConn domainB defSearchPolicy Nothing),
+        createFedConn domainB (FedConn domainA defSearchPolicy Nothing)
+      ]
+    void $ updateFedConn domainA domainB (FedConn domainB defSearchPolicy $ Just [])
```
Comment (battermann) [inline integration/test/Test/Connection.hs:317]: Why not create the policy the way we want it right away instead of first creating it and then updating it?

File: integration/test/Test/Connection.hs [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+  withFederatingBackendsAllowDynamic $ \(domainA, domainB, _) -> do
+    sequence_
+      [ createFedConn domainA (FedConn domainB defSearchPolicy Nothing),
+        createFedConn domainB (FedConn domainA defSearchPolicy Nothing)
+      ]
+    alice <- randomUser domainA def
+    bob <- randomUser domainB def {API.BrigInternal.team = True}
+
+    -- Alice's backend federates with Bob's team
+    void $ updateFedConn domainA domainB (FedConn domainB defSearchPolicy $ Just [])
```
Comment (battermann) [inline integration/test/Test/Connection.hs:337]: As above I would prefer to create the fedration policy the way we want right away.

File: libs/wire-api/src/Wire/API/Routes/Public/Brig.hs [hunk 6]
```diff
@@ -971,6 +971,7 @@ type ConnectionAPI =
     :<|> Named
            "create-connection"
            ( Summary "Create a connection to another user"
+               :> MakesFederatedCall 'Brig "get-users-by-ids"
```
Comment (battermann) [inline libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:974]: Should `CanThrow TeamsNotFederating` be added here?

File: services/brig/test/integration/API/User/Connection.hs [hunk 7]
```diff
@@ -62,7 +60,7 @@ tests ::
   FedClient 'Galley ->
   DB.ClientState ->
   TestTree
-tests cl _at opts p b _c g fedBrigClient fedGalleyClient db =
+tests cl _at p b _c g fedBrigClient _fedGalleyClient db =
```
Comment (battermann) [inline services/brig/test/integration/API/User/Connection.hs:63]: remove `_fedGalleyClient`?

File: services/brig/test/integration/Federation/End2end.hs [hunk 8]
```diff
@@ -592,7 +592,15 @@ claimRemoteKeyPackages brig1 brig2 = do
     Set.map (\e -> (e.user, e.client)) bundle.entries
       @?= Set.fromList [(bob, c) | c <- bobClients]
 
-testRemoteTypingIndicator :: Brig -> Brig -> Galley -> Galley -> Cannon -> Cannon -> Http ()
+testRemoteTypingIndicator ::
```
Comment (battermann) [inline services/brig/test/integration/Federation/End2end.hs:595]: This seems unrelated to this PR.

**Review by battermann — APPROVED** [review]
- Awesome that some tests from the old integration suite were ported!
- I would suggest to add tests for non-team users, too
- Looks good otherwise, only very marginal comments

File: integration/test/SetupHelpers.hs [hunk 9]
```diff
@@ -194,3 +194,54 @@ withFederatingBackendsAllowDynamic k = do
...
+-- | Fetch the one-to-one conversation between the two users that is in one of
+-- two possible states.
+getOne2OneConversation :: HasCallStack => Value -> Value -> One2OneConvState -> App Value
+getOne2OneConversation user1 user2 cnvState = do
+  l <- getAllConvs user1
+  let isWith users c = do
+        -- The conversation type 2 is for 1-to-1 conversations. Type 3 is for
+        -- the connection conversation, which is the state of the conversation
+        -- before the connection is fully established.
+        t <- (==) <$> (c %. "type" & asInt) <*> pure (toConvType cnvState)
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply integration/test/SetupHelpers.hs:243]: I copy-pasted this from the `Conversation` module where I had it.

To get the `type` field out of `c`, you need to run in the `App` monad that we have for tests.

File: integration/test/Test/Connection.hs (hunk 4 above)
Comment (mdimjasevic) [reply integration/test/Test/Connection.hs:317]: I forgot the details, but the type wrangling seemed a bit challenging for me to do it right away, hence I did it this way. A concrete suggestion on how to change this is more than welcome so I finally learn to do it nicely!

File: libs/wire-api/src/Wire/API/Routes/Public/Brig.hs (hunk 6 above)
Comment (mdimjasevic) [reply libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:974]: I don't think so. In Galley, we specify each individual error that a handler can throw and these are statically checked. In Brig, we have an ADT that a handler can throw, and these by the end of execution get converted to `StdError`, which is the code-label-message triple error response.

File: services/brig/test/integration/API/User/Connection.hs (hunk 7 above)
Comment (mdimjasevic) [reply services/brig/test/integration/API/User/Connection.hs:63]: I thought of that, but then left it (in case someone needs it). Now that I think about it again, we won't ever update these tests, so no one will re-introduce the need for that federation client; hence I'll remove it.

File: services/brig/test/integration/Federation/End2end.hs (hunk 8 above)
Comment (mdimjasevic) [reply services/brig/test/integration/Federation/End2end.hs:595]: This is a left-over from debugging this test as it was initially failing.

I tend to split the screen vertically and then these long signatures don't fit into a half of the screen, and it gets annoying having to move the cursor/slider to see what are the rest of the arguments.

If you insist, I'll remove it from this PR, but I'd argue the change makes the code more ergonomic.

File: integration/test/Test/Connection.hs (hunk 3 above)
Comment (MangoIV) [reply integration/test/Test/Connection.hs:245]: Iirc this does more than I'm allowed to do here. (To make the test replicate the old behaviour)

File: integration/test/SetupHelpers.hs (hunk 9 above)
[the remark this answers is not in the document]
Comment (battermann) [reply integration/test/SetupHelpers.hs:243]: Yeah, I saw later that it was just moved here. Anyway, something like this `t <- (== toConvType cnvState) <$> (c %. "type" & asInt)` should work as well.

File: libs/wire-api/src/Wire/API/Routes/Public/Brig.hs (hunk 6 above)
Comment (battermann) [reply libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:974]: I think it is useful for clients to know which status and labels to expect so that they can handle them.

File: services/brig/test/integration/Federation/End2end.hs (hunk 8 above)
Comment (battermann) [reply services/brig/test/integration/Federation/End2end.hs:595]: All good.

File: libs/wire-api/src/Wire/API/Routes/Public/Brig.hs (hunk 6 above)
Comment (mdimjasevic) [reply libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:974]: I agree, but with Brig we're not there (yet). Brig is in a half-finished state when it comes to migrating to Polysemy. With Polysemy we have the ability to list only specific errors an endpoint can return and have them checked by the compiler (as done in Galley). If I manually add this annotation to the endpoint, it will not be requested by the endpoint handler, and then we're back to manually annotating Swagger with errors an endpoint can return; such documentation is out of sync and out of date in no time.

Or am I getting this wrong?

Comment (MangoIV) [reply libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:974]: I think the ADT is specification enough, everything else has too much overhead.

Comment (MangoIV) [reply libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:974]: I'm not even aware; is that ADT somewhere added to Swagger?

File: integration/test/SetupHelpers.hs (hunk 9 above)
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply integration/test/SetupHelpers.hs]: I implemented it as you suggested. It's easier to read.

File: libs/wire-api/src/Wire/API/Routes/Public/Brig.hs (hunk 6 above)
Comment (mdimjasevic) [reply libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:974]: No. These ADTs are not reflected in Swagger. In this case, we have the `ConnectionError` error type, but before the endpoint returns, this error that is passed around within the `ConnectionM r` monad (just an `ExceptT`), is eventually converted to an `StdError`, e.g.:

```haskell
API.createConnection lself conn target !>> connError
```
where:
```haskell
connError :: ConnectionError -> Error
```

File: integration/test/Test/Connection.hs (hunk 4 above)
Comment (battermann) [reply integration/test/Test/Connection.hs:317]: I meant change it to just this:

```
    sequence_
      [ createFedConn domainA (FedConn domainB defSearchPolicy $ Just []),
        createFedConn domainB (FedConn domainA defSearchPolicy Nothing)
      ]
```
